netket/models/slater.py

```diff
@@ -0,0 +1,112 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import flax.linen as nn
+import jax.numpy as jnp
+
+from functools import partial
+
+from netket.experimental.hilbert import SpinOrbitalFermions
+from netket.utils.types import NNInitFunc, DType
+from netket.nn.masked_linear import default_kernel_init
+from netket import jax as nkjax
+
+
+def _log_det(A):
+    """Complex-aware logdeterminant built on top of
+    {func}`jax.numpy.linalg.slogdet` and combining the sign again.
+    """
+    sign, logabsdet = jnp.linalg.slogdet(A)
+    cplx_type = nkjax.dtype_complex(A.dtype)
+    return logabsdet.astype(cplx_type) + jnp.log(sign.astype(cplx_type))
+
+
+class LogSlater2nd(nn.Module):
+    r"""
+    A slater determinant ansatz for second-quantised spinless or spin-full
+    fermions.
+
+    The log-determinants of every spin sub-sector are summed together.
+
+    The total number of parameters of this ansatz is :math:`n_{\mathrm{o}} \times n_{\mathrm{f}}`,
+    where :math:`n_{\mathrm{o}}` is the number of orbitals and :math:`n_{\mathrm{f}}` is the
+    total number of fermions in the system.
+    """
+
+    hilbert: SpinOrbitalFermions
+    """The Hilbert space upon which this ansatz is defined. Used to determine the number of orbitals
+    and spin subspectors."""
+    kernel_init: NNInitFunc = default_kernel_init
+    """Initializer for the orbital parameters."""
+    param_dtype: DType = float
+    """Dtype of the orbital amplitudes."""
+
+    def __post_init__(self):
+        if not isinstance(self.hilbert, SpinOrbitalFermions):
+            raise TypeError(
+                "LogSlater2nd only supports 2nd quantised fermionic hilbert spaces."
+            )
+        if self.hilbert.n_fermions is None:
+            raise TypeError(
+                "LogSlater2nd only supports hilbert spaces with a "
+                "fixed number of fermions."
+            )
+        super().__post_init__()
+
+    def setup(self):
+        # Every determinant is a matrix of shape (n_orbitals, n_fermions_i) where
+        # n_fermions_i is the number of fermions in the i-th spin sector.
+        self.determinants = [
+            self.param(
+                f"M_{i}",
+                self.kernel_init,
+                (self.hilbert.n_orbitals, nf_i),
+                self.param_dtype,
+            )
+            for i, nf_i in enumerate(self.hilbert.n_fermions_per_spin)
+        ]
+
+    def __call__(self, n):
```

Comment:
Suggested alternative approach: just jnp.split the occupation numbers and define a function that selects the colums and produces the logdet.

Reply:
Nit: I'm selecting rows now, as it's epsilon more efficient in python and I think makes more sense. 

Yes, I agree with you that it would simplify the code.
 
My reasoning was that this approach saves us from splitting the samples, which would basically amount to copying the full samples while striding. For many sectors this would be inefficient. 
You can see it in this ansatz. Of course as soon as you add a bacflow it does not matter anymore.